### DISCUSSION TITLE
Spreading HTML element props in Astro components

### DISCUSSION THREAD

**Discussion by jimpala** [withastro/roadmap#848] [Proposal]
### Body

# Summary

Spreading props to HTML intrinsic elements in TypeScript appears to be a bit difficult at the moment - this might be helped by exporting the relevant types from Astro.

# Background & Motivation

Hi! Forgive me if this is misconceived — I may just be thinking too much in React!

Say I’m wanting to make an adapted `<input>` component `<InputText>`. Here’s the amended free example from Tailwind UI:

```html
<input type="text" class="block w-full rounded-md border-0 py-1.5 pl-7 pr-20 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6" placeholder="0.00">
```

Some of these attributes we may want to have default values that are already set, for example just `type` in this trivial example.

We might want to create a component in `InputText.astro` like so, which works analogously to an `<input>` element:

```astro
---
interface Props extends HTMLInputElement {
	foo: string;
}

const { foo, ...rest } = Astro.props;
---

<input
	type="text" class="block w-full blah"
	{...rest}
/>
```

This produces a couple of errors, the first being that the `type` and  `class` props of `<input>` will be overwritten. Amending for this using `Omit` and using the documented way of merging classes using `class:list` :

```astro
---
interface Props extends Omit<HTMLInputElement, "type"> {
	foo: string;
}

const { foo, class: className = "", ...rest } = Astro.props;
---

<input
	type="text" class={["block w-full blah", className]}
	{...rest}
/>
```

The error now is that the DOM `HTMLInputElement` is not compatible with Astro’s `<input>` element (this crops up for props like `form` and `list`).

I’ve had a look and it seems that the the types for these primitive elements are defined in Astro’s `DefinedIntrinsicElements` interface, but this isn’t exported.

Am I overthinking this or does this make it difficult to componentize elements tied to HTML primitive elements that can be flexibly used (like converting components like those in Tailwind UI into dedicated Astro components)? Or am I approaching this in totally the wrong way? Thanks!

# Goals

- Potentially expose Astro's type definitions like `DefinedIntrinsicElements` to facilitate easier props passing.

**Comment by delucis** [comment]
If I understood correctly, you might be looking for Astro's `HTMLAttributes` type? https://docs.astro.build/en/guides/typescript/#built-in-html-attributes

**Comment by jimpala** [comment]
Yep this is perfect, don't know how I missed it! Thank you @delucis 😄
